db/version_set.h

```diff
@@ -699,12 +699,28 @@ class VersionSet {
     return last_sequence_.load(std::memory_order_acquire);
   }
 
+  uint64_t LastToBeWrittenSequence() const {
+    return last_to_be_written_sequence_.load(std::memory_order_acquire);
+  }
+
   // Set the last sequence number to s.
   void SetLastSequence(uint64_t s) {
     assert(s >= last_sequence_);
+    // Last visible seqeunce must always be less than last written seq
+    assert(!db_options_->concurrent_wal_writes ||
+           s <= last_to_be_written_sequence_);
     last_sequence_.store(s, std::memory_order_release);
   }
 
+  void SetLastToBeWrittenSequence(uint64_t s) {
+    assert(s >= last_to_be_written_sequence_);
+    last_to_be_written_sequence_.store(s, std::memory_order_release);
+  }
+
+  uint64_t FetchAddLastToBeWrittenSequence(uint64_t s) {
```

Comment:
I ended up added an extra last_to_be_written_sequence_ that has to be set properly all over the code base: initialization, file injection, ... But PipelinedWriteImpl contains the complexity inside itself and does not expose existence of this separate seq number to the reset of code base.